geopandas/geodataframe.py

```diff
@@ -459,33 +459,28 @@ def dissolve(self, by=None, aggfunc='first'):
         """
 
         # Process non-spatial component
-        data = self.drop(labels=self.geometry.name, axis=1).copy()
+        data = self.drop(labels=self.geometry.name, axis=1)
         aggregated_data = data.groupby(by=by).agg(aggfunc)
 
 
         # Process spatial component
-        groupby_plus_geometry_cols = [self.geometry.name]
-        groupby_plus_geometry_cols.append(by)
-        geometry = self[groupby_plus_geometry_cols].copy()
-
         def merge_geometries(block):
-
             merged_geom = block.unary_union
+            return merged_geom
 
-            new_index = block.drop(self.geometry.name, axis=1).iloc[0][by]
-            merged_w_index = GeoSeries(merged_geom, index=Index(Series(new_index),name=by), 
-                                          name=self.geometry.name)
-            return merged_w_index
-
+        g = self.groupby(by=by, group_keys=False)[self.geometry.name].apply(merge_geometries)
 
-        g = geometry.groupby(by=by, group_keys=False).apply(merge_geometries)
-
-        aggregated_geometry = GeoDataFrame(g, 
+        # Aggregate
+        aggregated_geometry = GeoDataFrame(g,
                                            index=g.index,
```

Comment:
Is this needed? (not sure, just a question, I would think not, but if it is needed this is maybe also a bug/missing feature)